Error while checking build/O2/fullCI for 45b43464f6da8f0a5d653c1fe0e84d6303fd825a at 2023-10-26 08:44:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile reco_NOGPU.log
reco_NOGPU.log:[81896:calib-emcalchannel-badchannel]: [06:44:18][FATAL] error while setting up workflow in o2-calibration-emcal-channel-calib-workflow: filesystem error: cannot create directories: Permission denied [/scratch/services/detector_tmp/emc_calib]
[81890:BadMapCalibSpec]: [06:44:08][ERROR] Insufficient statistics: 7 entries in lowE histo, do nothing
[81896:calib-emcalchannel-badchannel]: [06:44:18][FATAL] error while setting up workflow in o2-calibration-emcal-channel-calib-workflow: filesystem error: cannot create directories: Permission denied [/scratch/services/detector_tmp/emc_calib]
[ERROR] pid 81896 (calib-emcalchannel-badchannel) crashed with or was killed with exit code 1


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12127/45b43464f6da8f0a5d653c1fe0e84d6303fd825a/build_O2_fullCI/pretty.html).

--- a/DataFormats/common/include/CommonDataFormat/InteractionRecord.h
+++ b/DataFormats/common/include/CommonDataFormat/InteractionRecord.h
@@ -109,7 +109,7 @@ struct InteractionRecord {
 
   float differenceInBCMUS(const InteractionRecord& other) const
   {
-    // return difference in bunch-crossings in ms
+    // return difference in bunch-crossings in us
     return differenceInBC(other) * o2::constants::lhc::LHCBunchSpacingMUS;
   }
 

--- a/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/include/TPCInterpolationWorkflow/TPCInterpolationSpec.h
+++ b/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/include/TPCInterpolationWorkflow/TPCInterpolationSpec.h
@@ -38,7 +38,7 @@ namespace tpc
 class TPCInterpolationDPL : public Task
 {
  public:
-  TPCInterpolationDPL(std::shared_ptr<o2::globaltracking::DataRequest> dr, std::shared_ptr<o2::base::GRPGeomRequest> gr, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput) : mDataRequest(dr), mGGCCDBRequest(gr), mUseMC(useMC), mProcessITSTPConly(processITSTPConly), mSendTrackData(sendTrackData), mDebugOutput(debugOutput) {}
+  TPCInterpolationDPL(std::shared_ptr<o2::globaltracking::DataRequest> dr, o2::dataformats::GlobalTrackID::mask_t src, std::shared_ptr<o2::base::GRPGeomRequest> gr, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput) : mDataRequest(dr), mSources(src), mGGCCDBRequest(gr), mUseMC(useMC), mProcessITSTPConly(processITSTPConly), mSendTrackData(sendTrackData), mDebugOutput(debugOutput) {}
   ~TPCInterpolationDPL() override = default;
   void init(InitContext& ic) final;
   void run(ProcessingContext& pc) final;
@@ -52,6 +52,7 @@ class TPCInterpolationDPL : public Task
   std::shared_ptr<o2::base::GRPGeomRequest> mGGCCDBRequest;
   o2::tpc::VDriftHelper mTPCVDriftHelper{};
   const o2::itsmft::TopologyDictionary* mITSDict = nullptr; ///< cluster patterns dictionary
+  o2::dataformats::GlobalTrackID::mask_t mSources{};        ///< which input sources are configured
   bool mUseMC{false}; ///< MC flag
   bool mProcessITSTPConly{false}; ///< should also tracks without outer point (ITS-TPC only) be processed?
   bool mProcessSeeds{false};      ///< process not only most complete track, but also its shorter parts
@@ -63,7 +64,7 @@ class TPCInterpolationDPL : public Task
 };
 
 /// create a processor spec
-framework::DataProcessorSpec getTPCInterpolationSpec(o2::dataformats::GlobalTrackID::mask_t src, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput);
+framework::DataProcessorSpec getTPCInterpolationSpec(o2::dataformats::GlobalTrackID::mask_t srcCls, o2::dataformats::GlobalTrackID::mask_t srcVtx, o2::dataformats::GlobalTrackID::mask_t srcTrk, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput);
 
 } // namespace tpc
 } // namespace o2

--- a/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/TPCInterpolationSpec.cxx
+++ b/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/TPCInterpolationSpec.cxx
@@ -63,7 +63,11 @@ void TPCInterpolationDPL::updateTimeDependentParams(ProcessingContext& pc)
     initOnceDone = true;
     // other init-once stuff
     const auto& param = SpacePointsCalibConfParam::Instance();
-    mInterpolation.init();
+    mInterpolation.init(mSources);
+    if (mProcessITSTPConly) {
+      mInterpolation.setProcessITSTPConly();
+    }
+    mInterpolation.setSqrtS(o2::base::GRPGeomHelper::instance().getGRPLHCIF()->getSqrtS());
     int nTfs = mSlotLength / (o2::base::GRPGeomHelper::getNHBFPerTF() * o2::constants::lhc::LHCOrbitMUS * 1e-6);
     bool limitTracks = (param.maxTracksPerCalibSlot < 0) ? false : true;
     int nTracksPerTfMax = (nTfs > 0 && limitTracks) ? param.maxTracksPerCalibSlot / nTfs : -1;
@@ -119,107 +123,15 @@ void TPCInterpolationDPL::finaliseCCDB(ConcreteDataMatcher& matcher, void* obj)
 
 void TPCInterpolationDPL::run(ProcessingContext& pc)
 {
-  LOG(info) << "TPC Interpolation Workflow initialized. Start processing...";
   mTimer.Start(false);
   RecoContainer recoData;
   recoData.collectData(pc, *mDataRequest.get());
   updateTimeDependentParams(pc);
-  const auto& param = SpacePointsCalibConfParam::Instance();
-
-  // load the input tracks
-  std::vector<o2::globaltracking::RecoContainer::GlobalIDSet> gidTables;
-  std::vector<o2::track::TrackParCov> seeds;
-  std::vector<float> trkTimes;
-  std::vector<GTrackID> gids;
-  std::unordered_map<int, int> trkCounters;
-  // make sure the map has entries for every possible track input type
-  trkCounters.insert(std::make_pair<int, int>(GTrackID::Source::ITSTPCTRDTOF, 0));
-  trkCounters.insert(std::make_pair<int, int>(GTrackID::Source::ITSTPCTRD, 0));
-  trkCounters.insert(std::make_pair<int, int>(GTrackID::Source::ITSTPCTOF, 0));
-  trkCounters.insert(std::make_pair<int, int>(GTrackID::Source::ITSTPC, 0));
-
-  // so that the flags can be used inside the lambda
-  bool processITSTPConly = mProcessITSTPConly;
-  bool processSeeds = mProcessSeeds;
-  // the creator goes from most complete track (ITS-TPC-TRD-TOF) to least complete one (ITS-TPC)
-  auto creator = [&gidTables, &seeds, &trkTimes, &recoData, &processITSTPConly, &processSeeds, &gids, &param, &trkCounters](auto& _tr, GTrackID _origID, float t0, float tErr) {
-    if constexpr (std::is_base_of_v<o2::track::TrackParCov, std::decay_t<decltype(_tr)>>) {
-      bool trackGood = true;
-      bool hasOuterPoint = false;
-      auto gidTable = recoData.getSingleDetectorRefs(_origID);
-      if (!gidTable[GTrackID::ITS].isIndexSet() || !gidTable[GTrackID::TPC].isIndexSet()) {
-        // ITS and TPC track is always needed. At this stage ITS afterburner tracks are also rejected
-        return true;
-      }
-      if (gidTable[GTrackID::TRD].isIndexSet()) {
-        // TRD specific cuts
-        const auto& trdTrk = recoData.getITSTPCTRDTrack<o2::trd::TrackTRD>(gidTable[GTrackID::ITSTPCTRD]);
-        if (trdTrk.getNtracklets() < param.minTRDNTrklts) {
-          trackGood = false;
-        }
-        hasOuterPoint = true;
-      }
-      if (gidTable[GTrackID::TOF].isIndexSet()) {
-        // TOF specific cuts (if any)
-        hasOuterPoint = true;
-      }
-      const auto itstpcTrk = &recoData.getTPCITSTrack(gidTable[GTrackID::ITSTPC]);
-      const auto itsTrk = &recoData.getITSTrack(gidTable[GTrackID::ITS]);
-      const auto tpcTrk = &recoData.getTPCTrack(gidTable[GTrackID::TPC]);
-      // apply track quality cuts
-      if (itsTrk->getChi2() / itsTrk->getNumberOfClusters() > param.maxITSChi2 || tpcTrk->getChi2() / tpcTrk->getNClusterReferences() > param.maxTPCChi2) {
-        // reduced chi2 cut is the same for all track types
-        trackGood = false;
-      }
-      if (!hasOuterPoint) {
-        // ITS-TPC track (does not have outer points in TRD or TOF)
-        if (!processITSTPConly) {
-          return true;
-        }
-        if (itsTrk->getNumberOfClusters() < param.minITSNClsNoOuterPoint || tpcTrk->getNClusterReferences() < param.minTPCNClsNoOuterPoint) {
-          trackGood = false;
-        }
-        if (itsTrk->getPt() < param.minPtNoOuterPoint) {
-          trackGood = false;
-        }
-      } else {
-        if (itsTrk->getNumberOfClusters() < param.minITSNCls || tpcTrk->getNClusterReferences() < param.minTPCNCls) {
-          trackGood = false;
-        }
-      }
-      if (trackGood) {
-        trkTimes.push_back(t0);
-        seeds.emplace_back(itsTrk->getParamOut()); // FIXME: should this not be a refit of the ITS track?
-        gidTables.emplace_back(gidTable);
-        gids.push_back(_origID);
-        trkCounters[_origID.getSource()] += 1;
-      }
-      if (processSeeds || (gidTable[GTrackID::TRD].isIndexSet() && gidTable[GTrackID::TOF].isIndexSet())) {
-        // for ITS-TPC-TRD-TOF tracks we are interested also in the ITS-TPC-TRD part
-        // we want to have both available, so return false for the full barrell tracks
-        return false;
-      } else {
-        return true;
-      }
-    } else {
-      return false;
-    }
-  };
-  recoData.createTracksVariadic(creator); // create track sample considered for interpolation
-  LOGP(info, "Created {} seeds. {} ITS-TPC-TRD-TOF, {} ITS-TPC-TRD, {} ITS-TPC-TOF, {} ITS-TPC",
-       seeds.size(), trkCounters.at(GTrackID::Source::ITSTPCTRDTOF), trkCounters.at(GTrackID::Source::ITSTPCTRD),
-       trkCounters.at(GTrackID::Source::ITSTPCTOF), trkCounters.at(GTrackID::Source::ITSTPC));
-
-  if (mUseMC) {
-    // possibly MC labels will be used to check filtering procedure performance before interpolation
-    // not yet implemented
-  }
-
-  mInterpolation.process(recoData, gids, gidTables, seeds, trkTimes, trkCounters);
+  mInterpolation.prepareInputTrackSample(recoData);
+  mInterpolation.process();
   mTimer.Stop();
   LOGF(info, "TPC interpolation timing: Cpu: %.3e Real: %.3e s", mTimer.CpuTime(), mTimer.RealTime());
-
-  if (param.writeUnfiltered) {
+  if (SpacePointsCalibConfParam::Instance().writeUnfiltered) {
     // these are the residuals and tracks before outlier rejection; they are not used in production
     pc.outputs().snapshot(Output{"GLO", "TPCINT_RES", 0, Lifetime::Timeframe}, mInterpolation.getClusterResidualsUnfiltered());
     if (mSendTrackData) {
@@ -234,7 +146,6 @@ void TPCInterpolationDPL::run(ProcessingContext& pc)
   if (mDebugOutput) {
     pc.outputs().snapshot(Output{"GLO", "TRKDATAEXT", 0, Lifetime::Timeframe}, mInterpolation.getTrackDataExtended());
   }
-
   mInterpolation.reset();
 }
 
@@ -244,7 +155,7 @@ void TPCInterpolationDPL::endOfStream(EndOfStreamContext& ec)
        mTimer.CpuTime(), mTimer.RealTime(), mTimer.Counter() - 1);
 }
 
-DataProcessorSpec getTPCInterpolationSpec(GTrackID::mask_t src, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput)
+DataProcessorSpec getTPCInterpolationSpec(GTrackID::mask_t srcCls, GTrackID::mask_t srcVtx, GTrackID::mask_t srcTrk, bool useMC, bool processITSTPConly, bool sendTrackData, bool debugOutput)
 {
   auto dataRequest = std::make_shared<DataRequest>();
   std::vector<OutputSpec> outputs;
@@ -253,12 +164,13 @@ DataProcessorSpec getTPCInterpolationSpec(GTrackID::mask_t src, bool useMC, bool
     LOG(fatal) << "MC usage must be disabled for this workflow, since it is not yet implemented";
   }
 
-  dataRequest->requestTracks(src, useMC);
-  dataRequest->requestClusters(src, useMC);
+  dataRequest->requestTracks(srcVtx, useMC);
+  dataRequest->requestClusters(srcCls, useMC);
+  dataRequest->requestPrimaryVertertices(useMC);
 
   auto ggRequest = std::make_shared<o2::base::GRPGeomRequest>(false,                             // orbitResetTime
                                                               true,                              // GRPECS=true
-                                                              false,                             // GRPLHCIF
+                                                              true,                              // GRPLHCIF
                                                               true,                              // GRPMagField
                                                               true,                              // askMatLUT
                                                               o2::base::GRPGeomRequest::Aligned, // geometry
@@ -284,7 +196,7 @@ DataProcessorSpec getTPCInterpolationSpec(GTrackID::mask_t src, bool useMC, bool
     "tpc-track-interpolation",
     dataRequest->inputs,
     outputs,
-    AlgorithmSpec{adaptFromTask<TPCInterpolationDPL>(dataRequest, ggRequest, useMC, processITSTPConly, sendTrackData, debugOutput)},
+    AlgorithmSpec{adaptFromTask<TPCInterpolationDPL>(dataRequest, srcTrk, ggRequest, useMC, processITSTPConly, sendTrackData, debugOutput)},
     Options{
       {"matCorrType", VariantType::Int, 2, {"material correction type (definition in Propagator.h)"}},
       {"sec-per-slot", VariantType::UInt32, 600u, {"number of seconds per calibration time slot (put 0 for infinite slot length)"}},

--- a/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/tpc-interpolation-workflow.cxx
+++ b/Detectors/GlobalTrackingWorkflow/tpcinterpolationworkflow/src/tpc-interpolation-workflow.cxx
@@ -37,8 +37,8 @@ void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
     {"disable-root-input", VariantType::Bool, false, {"disable root-files input readers"}},
     {"disable-root-output", VariantType::Bool, false, {"disable root-files output writers"}},
     {"disable-mc", VariantType::Bool, false, {"disable MC propagation even if available"}},
-    {"enable-itsonly", VariantType::Bool, false, {"process tracks without outer point (ITS-TPC only)"}},
-    {"tracking-sources", VariantType::String, std::string{GID::ALL}, {"comma-separated list of sources to use for tracking"}},
+    {"vtx-sources", VariantType::String, std::string{GID::ALL}, {"comma-separated list of sources used for the vertex finding"}},
+    {"tracking-sources", VariantType::String, std::string{GID::ALL}, {"comma-separated list of sources to use for track inter-/extrapolation"}},
     {"send-track-data", VariantType::Bool, false, {"Send also the track information to the aggregator"}},
     {"debug-output", VariantType::Bool, false, {"Dump extended tracking information for debugging"}},
     {"configKeyValues", VariantType::String, "", {"Semicolon separated key=value strings ..."}}};
@@ -66,25 +66,41 @@ void customize(std::vector<o2::framework::CompletionPolicy>& policies)
 WorkflowSpec defineDataProcessing(ConfigContext const& configcontext)
 {
   WorkflowSpec specs;
-  GID::mask_t allowedSources = GID::getSourcesMask("ITS,TPC,TRD,TOF,ITS-TPC,ITS-TPC-TRD,ITS-TPC-TOF,ITS-TPC-TRD-TOF");
+  GID::mask_t allowedSources = GID::getSourcesMask("ITS-TPC,ITS-TPC-TRD,ITS-TPC-TOF,ITS-TPC-TRD-TOF");
+  GID::mask_t srcVtx = allowedSources & GID::getSourcesMask(configcontext.options().get<std::string>("vtx-sources"));
+  GID::mask_t srcTracks = allowedSources & GID::getSourcesMask(configcontext.options().get<std::string>("tracking-sources"));
+  if (srcTracks.count() > srcVtx.count()) {
+    LOGP(error, "More sources configured for inter-/extrapolation: {} than for vertexing: {}. Additional sources will be ignored", GID::getSourcesNames(srcTracks), GID::getSourcesNames(srcVtx));
+    srcTracks &= srcVtx;
+  }
+  LOG(debug) << "Data sources for inter-/extrapolation: " << GID::getSourcesNames(srcTracks);
+  // check first if ITS-TPC tracks were specifically requested from command line
+  bool processITSTPConly = srcTracks[GID::ITSTPC];
+  srcTracks |= GID::getSourcesMask("ITS,TPC,ITS-TPC"); // now add them in any case
+  srcVtx |= srcTracks;
+  GID::mask_t srcClusters = srcTracks;
+  if (srcTracks[GID::ITSTPCTRD] || srcTracks[GID::ITSTPCTRDTOF]) {
+    srcClusters |= GID::getSourcesMask("TRD");
+  }
+  if (srcTracks[GID::ITSTPCTOF] || srcTracks[GID::ITSTPCTRDTOF]) {
+    srcClusters |= GID::getSourcesMask("TOF");
+  }
   // Update the (declared) parameters if changed from the command line
   o2::conf::ConfigurableParam::updateFromString(configcontext.options().get<std::string>("configKeyValues"));
   // write the configuration used for the workflow
   o2::conf::ConfigurableParam::writeINI("o2tpcinterpolation-workflow_configuration.ini");
   auto useMC = !configcontext.options().get<bool>("disable-mc");
   useMC = false; // force disabling MC as long as it is not implemented
-  auto processITSTPConly = configcontext.options().get<bool>("enable-itsonly");
   auto sendTrackData = configcontext.options().get<bool>("send-track-data");
   auto debugOutput = configcontext.options().get<bool>("debug-output");
-  GID::mask_t src = allowedSources & GID::getSourcesMask(configcontext.options().get<std::string>("tracking-sources"));
-  LOG(info) << "Data sources: " << GID::getSourcesNames(src);
 
-  specs.emplace_back(o2::tpc::getTPCInterpolationSpec(src, useMC, processITSTPConly, sendTrackData, debugOutput));
+  specs.emplace_back(o2::tpc::getTPCInterpolationSpec(srcClusters, srcVtx, srcTracks, useMC, processITSTPConly, sendTrackData, debugOutput));
   if (!configcontext.options().get<bool>("disable-root-output")) {
     specs.emplace_back(o2::tpc::getTPCResidualWriterSpec(sendTrackData, debugOutput));
   }
 
-  o2::globaltracking::InputHelper::addInputSpecs(configcontext, specs, src, src, src, useMC);
+  o2::globaltracking::InputHelper::addInputSpecs(configcontext, specs, srcClusters, srcVtx, srcVtx, useMC);
+  o2::globaltracking::InputHelper::addInputSpecsPVertex(configcontext, specs, useMC); // P-vertex is always needed
 
   // configure dpl timer to inject correct firstTForbit: start from the 1st orbit of TF containing 1st sampled orbit
   o2::raw::HBFUtilsInitializer hbfIni(configcontext, specs);

--- a/Detectors/TPC/calibration/SpacePoints/include/SpacePoints/SpacePointsCalibConfParam.h
+++ b/Detectors/TPC/calibration/SpacePoints/include/SpacePoints/SpacePointsCalibConfParam.h
@@ -38,6 +38,9 @@ struct SpacePointsCalibConfParam : public o2::conf::ConfigurableParamHelper<Spac
   float maxITSChi2 = 20.f;         ///< cut on ITS reduced chi2
   float maxTRDChi2 = 10.f;         ///< cut on TRD reduced chi2
   float minPtNoOuterPoint = 0.8f;  ///< minimum pt for ITS-TPC tracks to be considered for extrapolation
+  bool ignoreNonPVContrib = true;  ///< flag if tracks which did not contribute to the PV should be ignored or not
+  bool enableTrackDownsampling = false; ///< flag if track sampling shall be enabled or not
+  float tsalisThreshold = 1.f;          ///< in case the sampling functions returns a value smaller than this the track is discarded (1. means no downsampling)
 
   // other settings for track interpolation
   float sigYZ2TOF{.75f}; ///< for now assume cluster error for TOF equal for all clusters in both Y and Z

--- a/Detectors/TPC/calibration/SpacePoints/include/SpacePoints/TrackInterpolation.h
+++ b/Detectors/TPC/calibration/SpacePoints/include/SpacePoints/TrackInterpolation.h
@@ -99,6 +99,9 @@ struct TrackDataCompact {
   ClassDefNV(TrackDataCompact, 1);
 };
 
+// TODO add to UnbinnedResid::sec flag if cluster was used or not
+// TODO add to TrackDataCompact::sourceId flag if track was used or not (if possible)
+
 /// Heavy structure with track parameterizations + track points for debugging
 struct TrackDataExtended {
   o2::dataformats::GlobalTrackID gid{};              ///< GID of the most global barrel track
@@ -185,11 +188,23 @@ class TrackInterpolation
 
   // -------------------------------------- processing functions --------------------------------------------------
 
-  /// Initialize everything
-  void init();
+  /// Initialize everything, set the requested track sources
+  void init(o2::dataformats::GlobalTrackID::mask_t src);
+
+  /// Check if input track passes configured cuts
+  bool isInputTrackAccepted(const o2::dataformats::GlobalTrackID& gid, const o2::globaltracking::RecoContainer::GlobalIDSet& gidTable, const o2::dataformats::PrimaryVertex& pv) const;
+
+  /// For given vertex track source which is not in mSourcesConfigured find the seeding source which is enabled
+  o2::dataformats::GlobalTrackID::Source findValidSource(o2::dataformats::GlobalTrackID::Source src) const;
+
+  /// Prepare input track sample (not relying on CreateTracksVariadic functionality)
+  void prepareInputTrackSample(const o2::globaltracking::RecoContainer& inp);
+
+  /// Given the defined downsampling factor tsalisThreshold check if track is selected
+  bool isTrackSelected(const o2::track::TrackParCov& trk) const;
 
   /// Main processing function
-  void process(const o2::globaltracking::RecoContainer& inp, const std::vector<o2::dataformats::GlobalTrackID>& gids, const std::vector<o2::globaltracking::RecoContainer::GlobalIDSet>& gidTables, std::vector<o2::track::TrackParCov>& seeds, const std::vector<float>& trkTimes, const std::unordered_map<int, int>& trkCounters);
+  void process();
 
   /// Extrapolate ITS-only track through TPC and store residuals to TPC clusters along the way
   /// \param seed index
@@ -249,6 +264,12 @@ class TrackInterpolation
   /// Enable processing of seeds
   void setProcessSeeds() { mProcessSeeds = true; }
 
+  /// Enable ITS-TPC only processing
+  void setProcessITSTPConly() { mProcessITSTPConly = true; }
+
+  /// Set the centre of mass energy required for pT downsampling Tsalis function
+  void setSqrtS(float s) { mSqrtS = s; }
+
   // --------------------------------- output ---------------------------------------------
   std::vector<UnbinnedResid>& getClusterResiduals() { return mClRes; }
   std::vector<TrackDataCompact>& getTrackDataCompact() { return mTrackDataCompact; }
@@ -266,18 +287,23 @@ class TrackInterpolation
   float mTPCVDrift = -1.;       ///< TPC drift speed in cm/microseconds
   float mTPCDriftTimeOffset = 0.;                    ///< TPC drift time bias in cm/mus
   float mTPCDriftTimeOffsetRef = 0.;                 ///< TPC nominal (e.g. at the start of run) drift time bias in cm/mus
+  float mSqrtS{13600.f};                             ///< centre of mass energy set from LHC IF
   MatCorrType mMatCorr{MatCorrType::USEMatCorrNONE}; ///< if material correction should be done
   int mMaxTracksPerTF{-1};                           ///< max number of tracks to be processed per TF (-1 means there is no limit)
   int mAddTracksITSTPC{0};                           ///< number of ITS-TPC tracks which can be processed in addition to mMaxTracksPerTF
   bool mDumpTrackPoints{false};                      ///< dump also track points in ITS, TRD and TOF
   bool mProcessSeeds{false};                         ///< in case for global tracks also their shorter parts are processed separately
+  bool mProcessITSTPConly{false};                    ///< flag, whether or not to extrapolate ITS-only through TPC
+  o2::dataformats::GlobalTrackID::mask_t mSourcesConfigured; ///< keep only the matches here, not the individual detector contributors
 
   // input
   const o2::globaltracking::RecoContainer* mRecoCont = nullptr;                            ///< input reco container
-  const std::vector<o2::dataformats::GlobalTrackID>* mGIDs = nullptr;                      ///< GIDs of input tracks
-  const std::vector<o2::globaltracking::RecoContainer::GlobalIDSet>* mGIDtables = nullptr; ///< GIDs of contributors from single detectors for each seed
-  const std::vector<float>* mTrackTimes = nullptr;                                         ///< time estimates for all input tracks in micro seconds
-  std::vector<o2::track::TrackParCov>* mSeeds = nullptr;                                   ///< seeding track parameters (ITS tracks)
+  std::vector<o2::dataformats::GlobalTrackID> mGIDs{};                                     ///< GIDs of input tracks
+  std::vector<o2::globaltracking::RecoContainer::GlobalIDSet> mGIDtables{};                ///< GIDs of contributors from single detectors for each seed
+  std::vector<float> mTrackTimes{};                                                        ///< time estimates for all input tracks in micro seconds
+  std::vector<o2::track::TrackParCov> mSeeds{};                                            ///< seeding track parameters (ITS tracks)
+  std::map<int, int> mTrackTypes;                                                          ///< mapping of track source to array index in mTrackIndices
+  std::array<std::vector<o2::dataformats::GlobalTrackID>, 4> mTrackIndices;                ///< keep GIDs of input tracks separately for each track type
   gsl::span<const TPCClRefElem> mTPCTracksClusIdx;                                         ///< input TPC cluster indices from span
   const ClusterNativeAccess* mTPCClusterIdxStruct = nullptr; ///< struct holding the TPC cluster indices
   // ITS specific input only needed for debugging

--- a/Detectors/TPC/calibration/SpacePoints/src/TrackInterpolation.cxx
+++ b/Detectors/TPC/calibration/SpacePoints/src/TrackInterpolation.cxx
@@ -23,10 +23,13 @@
 #include "DataFormatsTPC/Defs.h"
 #include "DataFormatsTRD/Constants.h"
 #include "DetectorsCommonDataFormats/DetID.h"
+#include "ReconstructionDataFormats/PrimaryVertex.h"
+#include "ReconstructionDataFormats/VtxTrackRef.h"
+#include "MathUtils/Tsallis.h"
 #include "TRDBase/PadPlane.h"
 #include "TMath.h"
 #include "DataFormatsTPC/VDriftCorrFact.h"
-#include <fairlogger/Logger.h>
+#include "Framework/Logger.h"
 #include <set>
 #include <algorithm>
 #include <random>
@@ -35,7 +38,7 @@ using namespace o2::tpc;
 using GTrackID = o2::dataformats::GlobalTrackID;
 using DetID = o2::detectors::DetID;
 
-void TrackInterpolation::init()
+void TrackInterpolation::init(o2::dataformats::GlobalTrackID::mask_t src)
 {
   // perform initialization
   LOG(info) << "Start initializing TrackInterpolation";
@@ -54,28 +57,165 @@ void TrackInterpolation::init()
   mGeoTRD = o2::trd::Geometry::instance();
   mParams = &SpacePointsCalibConfParam::Instance();
 
+  mSourcesConfigured = src;
+  mTrackTypes.insert({GTrackID::ITSTPC, 0});
+  mTrackTypes.insert({GTrackID::ITSTPCTRD, 1});
+  mTrackTypes.insert({GTrackID::ITSTPCTOF, 2});
+  mTrackTypes.insert({GTrackID::ITSTPCTRDTOF, 3});
+
   mInitDone = true;
-  LOG(info) << "Done initializing TrackInterpolation";
+  LOGP(info, "Done initializing TrackInterpolation. Configured track input: {}", GTrackID::getSourcesNames(mSourcesConfigured));
 }
 
-void TrackInterpolation::process(const o2::globaltracking::RecoContainer& inp, const std::vector<GTrackID>& gids, const std::vector<o2::globaltracking::RecoContainer::GlobalIDSet>& gidTables, std::vector<o2::track::TrackParCov>& seeds, const std::vector<float>& trkTimes, const std::unordered_map<int, int>& trkCounters)
+bool TrackInterpolation::isInputTrackAccepted(const GTrackID& gid, const o2::globaltracking::RecoContainer::GlobalIDSet& gidTable, const o2::dataformats::PrimaryVertex& pv) const
+{
+  LOGP(debug, "Check if input track {} is accepted", gid.asString());
+  bool hasOuterPoint = gidTable[GTrackID::TRD].isIndexSet() || gidTable[GTrackID::TOF].isIndexSet();
+  if (!hasOuterPoint && !mProcessITSTPConly) {
+    return false; // don't do ITS-only extrapolation through TPC
+  }
+  const auto itsTrk = mRecoCont->getITSTrack(gidTable[GTrackID::ITS]);
+  const auto tpcTrk = mRecoCont->getTPCTrack(gidTable[GTrackID::TPC]);
+
+  if (gidTable[GTrackID::TRD].isIndexSet()) {
+    // TRD specific cuts
+    const auto& trdTrk = mRecoCont->getITSTPCTRDTrack<o2::trd::TrackTRD>(gidTable[GTrackID::ITSTPCTRD]);
+    if (trdTrk.getNtracklets() < mParams->minTRDNTrklts) {
+      return false;
+    }
+  }
+  // reduced chi2 cut is the same for all track types
+  if (itsTrk.getChi2() / itsTrk.getNumberOfClusters() > mParams->maxITSChi2 || tpcTrk.getChi2() / tpcTrk.getNClusterReferences() > mParams->maxTPCChi2) {
+    return false;
+  }
+  if (!hasOuterPoint) {
+    // ITS-TPC track (does not have outer points in TRD or TOF)
+    if (itsTrk.getNumberOfClusters() < mParams->minITSNClsNoOuterPoint || tpcTrk.getNClusterReferences() < mParams->minTPCNClsNoOuterPoint) {
+      return false;
+    }
+    if (itsTrk.getPt() < mParams->minPtNoOuterPoint) {
+      return false;
+    }
+  } else {
+    if (itsTrk.getNumberOfClusters() < mParams->minITSNCls || tpcTrk.getNClusterReferences() < mParams->minTPCNCls) {
+      return false;
+    }
+  }
+
+  auto trc = mRecoCont->getTrackParam(gid);
+  o2::dataformats::DCA dca;
+  auto prop = o2::base::Propagator::Instance();
+  if (!prop->propagateToDCA(pv, trc, prop->getNominalBz(), 2., o2::base::PropagatorF::MatCorrType::USEMatCorrLUT, &dca)) {
+    return false;
+  }
+  if (dca.getR2() > mParams->maxDCA * mParams->maxDCA) {
+    return false;
+  }
+  return true;
+}
+
+GTrackID::Source TrackInterpolation::findValidSource(GTrackID::Source src) const
+{
+  LOGP(debug, "Trying to find valid source for {}", GTrackID::getSourcesNames(src));
+  if (src == GTrackID::ITSTPCTRDTOF) {
+    if (mSourcesConfigured[GTrackID::ITSTPCTRD]) {
+      return GTrackID::ITSTPCTRD;
+    } else if (mSourcesConfigured[GTrackID::ITSTPC]) {
+      return GTrackID::ITSTPC;
+    } else {
+      return GTrackID::NSources;
+    }
+  } else if (src == GTrackID::ITSTPCTRD || src == GTrackID::ITSTPCTOF) {
+    if (mSourcesConfigured[GTrackID::ITSTPC]) {
+      return GTrackID::ITSTPC;
+    } else {
+      return GTrackID::NSources;
+    }
+  } else {
+    return GTrackID::NSources;
+  }
+}
+
+void TrackInterpolation::prepareInputTrackSample(const o2::globaltracking::RecoContainer& inp)
+{
+  mRecoCont = &inp;
+  uint32_t nTrackSeeds = 0;
+  uint32_t countSeedCandidates[4] = {0};
+  auto pvvec = mRecoCont->getPrimaryVertices();
+  auto trackIndex = mRecoCont->getPrimaryVertexMatchedTracks(); // Global ID's for associated tracks
+  auto vtxRefs = mRecoCont->getPrimaryVertexMatchedTrackRefs(); // references from vertex to these track IDs
+  int nv = vtxRefs.size() - 1;
+  GTrackID::mask_t allowedSources = GTrackID::getSourcesMask("ITS-TPC,ITS-TPC-TRD,ITS-TPC-TOF,ITS-TPC-TRD-TOF");
+
+  for (int iv = 0; iv < nv; iv++) {
+    LOGP(debug, "processing PV {} of {}", iv, nv);
+
+    const auto& vtref = vtxRefs[iv];
+    auto pv = pvvec[iv];
+    for (int is = GTrackID::NSources; is >= 0; is--) {
+      if (!allowedSources[is]) {
+        continue;
+      }
+      LOGP(debug, "Checking source {}", is);
+      int idMin = vtref.getFirstEntryOfSource(is), idMax = idMin + vtref.getEntriesOfSource(is);
+      for (int i = idMin; i < idMax; i++) {
+        auto vid = trackIndex[i];
+        if (mParams->ignoreNonPVContrib && !vid.isPVContributor()) {
+          continue;
+        }
+        if (vid.isAmbiguous()) {
+          continue;
+        }
+        auto gidTable = mRecoCont->getSingleDetectorRefs(vid);
+        if (!mSourcesConfigured[is]) {
+          auto src = findValidSource(static_cast<GTrackID::Source>(vid.getSource()));
+          if (src == GTrackID::ITSTPCTRD || src == GTrackID::ITSTPC) {
+            LOGP(debug, "Found valid source {}", GTrackID::getSourcesNames(src));
+            vid = gidTable[src];
+            gidTable = mRecoCont->getSingleDetectorRefs(vid);
+          } else {
+            break; // no valid source for this vertex track source exists
+          }
+        }
+        ++countSeedCandidates[mTrackTypes[vid.getSource()]];
+        LOGP(debug, "Checking vid {}", vid.asString());
+        if (!isInputTrackAccepted(vid, gidTable, pv)) {
+          continue;
+        }
+        mSeeds.push_back(mRecoCont->getITSTrack(gidTable[GTrackID::ITS]).getParamOut());
+        mGIDs.push_back(vid);
+        mGIDtables.push_back(gidTable);
+        mTrackTimes.push_back(pv.getTimeStamp().getTimeStamp());
+        mTrackIndices[mTrackTypes[vid.getSource()]].push_back(nTrackSeeds++);
+      }
+    }
+  }
+
+  LOGP(info, "Created {} seeds. {} out of {} ITS-TPC-TRD-TOF, {} out of {} ITS-TPC-TRD, {} out of {} ITS-TPC-TOF, {} out of {} ITS-TPC",
+       nTrackSeeds,
+       mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRDTOF]].size(), countSeedCandidates[mTrackTypes[GTrackID::ITSTPCTRDTOF]],
+       mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRD]].size(), countSeedCandidates[mTrackTypes[GTrackID::ITSTPCTRD]],
+       mTrackIndices[mTrackTypes[GTrackID::ITSTPCTOF]].size(), countSeedCandidates[mTrackTypes[GTrackID::ITSTPCTOF]],
+       mTrackIndices[mTrackTypes[GTrackID::ITSTPC]].size(), countSeedCandidates[mTrackTypes[GTrackID::ITSTPC]]);
+}
+
+bool TrackInterpolation::isTrackSelected(const o2::track::TrackParCov& trk) const
+{
+  std::random_device rd;
+  std::mt19937 g(rd());
+  std::uniform_real_distribution<> distr(0., 1.);
+  float weight = 0;
+  return o2::math_utils::Tsallis::downsampleTsallisCharged(trk.getPt(), mParams->tsalisThreshold, mSqrtS, weight, distr(g), o2::constants::physics::MassPionCharged);
+}
+
+void TrackInterpolation::process()
 {
   // main processing function
-
   if (!mInitDone) {
     LOG(error) << "Initialization not yet done. Aborting...";
     return;
   }
-
-  // reset output vectors
-  reset();
-
   // set the input containers
-  mRecoCont = &inp;
-  mGIDs = &gids;
-  mGIDtables = &gidTables;
-  mSeeds = &seeds;
-  mTrackTimes = &trkTimes;
   mTPCTracksClusIdx = mRecoCont->getTPCTracksClusterRefs();
   mTPCClusterIdxStruct = &mRecoCont->getTPCClusters();
   if (mDumpTrackPoints) {
@@ -93,96 +233,78 @@ void TrackInterpolation::process(const o2::globaltracking::RecoContainer& inp, c
     o2::its::ioutils::convertCompactClusters(clusITS, pattIt, mITSClustersArray, mITSDict);
   }
 
-  int nSeeds = mSeeds->size();
-  mTrackData.reserve(nSeeds);
-  mClRes.reserve(nSeeds * param::NPadRows);
-
-  // to obtain ITS-TPC-TRD-TOF track from ITS-TPC-TRD track we fill the trkMap
-  std::unordered_map<GTrackID, int> trkMap;
-  for (int iTrk = 0; iTrk < trkCounters.at(GTrackID::Source::ITSTPCTRDTOF); ++iTrk) {
-    const auto& gidTable = (*mGIDtables)[iTrk];
-    trkMap.emplace(std::make_pair(gidTable[GTrackID::ITSTPCTRD], iTrk));
-  }
-
   // In case we have more input tracks available than are required per TF
   // we want to sample them. But we still prefer global ITS-TPC-TRD-TOF tracks
   // over ITS-TPC-TRD tracks and so on. So we have to shuffle the indices
   // in blocks.
-  // The input GIDs are sorted. ITS-TPC-TRD-TOF are first followed by ITS-TPC-TRD,
-  // ITS-TPC-TOF and ITS-TPC
   std::random_device rd;
   std::mt19937 g(rd());
-  std::vector<int> trackIndices(nSeeds);
-  std::iota(trackIndices.begin(), trackIndices.end(), 0);
+  std::vector<GTrackID> trackIndices; // here we keep the GIDs for all track types in a single vector to use in loop
+  std::shuffle(mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRDTOF]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRDTOF]].end(), g);
+  std::shuffle(mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRD]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRD]].end(), g);
+  std::shuffle(mTrackIndices[mTrackTypes[GTrackID::ITSTPCTOF]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTOF]].end(), g);
+  std::shuffle(mTrackIndices[mTrackTypes[GTrackID::ITSTPC]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPC]].end(), g);
+  trackIndices.insert(trackIndices.end(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRDTOF]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRDTOF]].end());
+  trackIndices.insert(trackIndices.end(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRD]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTRD]].end());
+  trackIndices.insert(trackIndices.end(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTOF]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPCTOF]].end());
+  trackIndices.insert(trackIndices.end(), mTrackIndices[mTrackTypes[GTrackID::ITSTPC]].begin(), mTrackIndices[mTrackTypes[GTrackID::ITSTPC]].end());
 
-  std::shuffle(trackIndices.begin(), trackIndices.begin() + trkCounters.at(GTrackID::Source::ITSTPCTRDTOF), g);
-  int nTracks = trkCounters.at(GTrackID::Source::ITSTPCTRDTOF);
-  std::shuffle(trackIndices.begin() + nTracks, trackIndices.begin() + nTracks + trkCounters.at(GTrackID::Source::ITSTPCTRD), g);
-  nTracks += trkCounters.at(GTrackID::Source::ITSTPCTRD);
-  std::shuffle(trackIndices.begin() + nTracks, trackIndices.begin() + nTracks + trkCounters.at(GTrackID::Source::ITSTPCTOF), g);
-  nTracks += trkCounters.at(GTrackID::Source::ITSTPCTOF);
-  std::shuffle(trackIndices.begin() + nTracks, trackIndices.begin() + nTracks + trkCounters.at(GTrackID::Source::ITSTPC), g);
-
-  std::vector<int> globalTracksToCheck;
+  int nSeeds = mSeeds.size();
+  int maxOutputTracks = (mMaxTracksPerTF >= 0) ? mMaxTracksPerTF + mAddTracksITSTPC : nSeeds;
+  mTrackData.reserve(maxOutputTracks);
+  mClRes.reserve(maxOutputTracks * param::NPadRows);
   bool maxTracksReached = false;
-  for (int iSeed = trkCounters.at(GTrackID::Source::ITSTPCTRDTOF); iSeed < nSeeds; ++iSeed) {
-    if (!maxTracksReached && mMaxTracksPerTF >= 0 && mTrackDataCompact.size() >= mMaxTracksPerTF) {
-      maxTracksReached = true;
-      // if mAddTracksITSTPC > 0 we will still process up to mAddTracksITSTPC tracks,
-      // otherwise the following if statement is also true and we break
-      if (trkCounters.at(GTrackID::Source::ITSTPC) > 0) {
-        iSeed = nSeeds - trkCounters.at(GTrackID::Source::ITSTPC);
-      } else {
-        // no ITS-TPC tracks available which could be processed
-        break;
-      }
-    }
+  for (int iSeed = 0; iSeed < nSeeds; ++iSeed) {
     if (mMaxTracksPerTF >= 0 && mTrackDataCompact.size() >= mMaxTracksPerTF + mAddTracksITSTPC) {
       LOG(info) << "Maximum number of tracks per TF reached. Skipping the remaining " << nSeeds - iSeed << " tracks.";
       break;
     }
-    if (auto search = trkMap.find((*mGIDs)[iSeed]); search != trkMap.end()) {
-      // for this ITS-TPC-TRD track we also have a match in TOF
-      if (mParams->debugTRDTOF) {
-        // process the ITS-TPC-TRD-TOF track later, in addition to the ITS-TPC-TRD track
-        globalTracksToCheck.push_back(search->second);
-      } else {
-        // process the ITS-TPC-TRD-TOF track and skip its ITS-TPC-TRD track seed
-        interpolateTrack(search->second);
-        continue;
+    int seedIndex = trackIndices[iSeed];
+    if (mParams->enableTrackDownsampling && !isTrackSelected(mSeeds[seedIndex])) {
+      continue;
+    }
+    if (mGIDs[seedIndex].includesDet(DetID::TRD) || mGIDs[seedIndex].includesDet(DetID::TOF)) {
+      interpolateTrack(seedIndex);
+      if (mProcessSeeds) {
+        if (mGIDs[seedIndex].includesDet(DetID::TRD) && mGIDs[seedIndex].includesDet(DetID::TOF)) {
+          mGIDs.push_back(mGIDtables[seedIndex][GTrackID::ITSTPCTRD]);
+          mGIDtables.push_back(mRecoCont->getSingleDetectorRefs(mGIDs.back()));
+          mTrackTimes.push_back(mTrackTimes[seedIndex]);
+          mSeeds.push_back(mSeeds[seedIndex]);
+        }
+        mGIDs.push_back(mGIDtables[seedIndex][GTrackID::ITSTPC]);
+        mGIDtables.push_back(mRecoCont->getSingleDetectorRefs(mGIDs.back()));
+        mTrackTimes.push_back(mTrackTimes[seedIndex]);
+        mSeeds.push_back(mSeeds[seedIndex]);
       }
-    }
-    if (gids[trackIndices[iSeed]].includesDet(DetID::TRD) || gids[trackIndices[iSeed]].includesDet(DetID::TOF)) {
-      interpolateTrack(trackIndices[iSeed]);
     } else {
-      extrapolateTrack(trackIndices[iSeed]);
+      extrapolateTrack(seedIndex);
     }
   }
-  // irrespective of the number of tracks already processed, interpolate the ITS-TPC-TRD-TOF tracks
-  // which belong to the ITS-TPC-TRD tracks that were already processed, to allow their analysis
-  // offline
-  // the globalTracksToCheck vector is only filled in case mParams->debugTRDTOF == true
-  LOGP(info, "Processing {} ITS-TPC-TRD-TOF tracks for which the ITS-TPC-TRD track was already done", globalTracksToCheck.size());
-  for (auto iTrk : globalTracksToCheck) {
-    interpolateTrack(iTrk);
+  for (int iSeed = nSeeds; iSeed < (int)mSeeds.size(); ++iSeed) {
+    // this loop will only be entered in case mProcessSeeds is set
+    if (mGIDs[iSeed].includesDet(DetID::TRD) || mGIDs[iSeed].includesDet(DetID::TOF)) {
+      interpolateTrack(iSeed);
+    } else {
+      extrapolateTrack(iSeed);
+    }
   }
-
   LOG(info) << "Could process " << mTrackData.size() << " tracks successfully";
 }
 
 void TrackInterpolation::interpolateTrack(int iSeed)
 {
-  LOGP(debug, "Starting track interpolation for seed {}", iSeed);
+  LOGP(debug, "Starting track interpolation for GID {}", mGIDs[iSeed].asString());
   TrackData trackData;
   std::unique_ptr<TrackDataExtended> trackDataExtended;
   std::vector<TPCClusterResiduals> clusterResiduals;
   auto propagator = o2::base::Propagator::Instance();
-  const auto& gidTable = (*mGIDtables)[iSeed];
+  const auto& gidTable = mGIDtables[iSeed];
   const auto& trkTPC = mRecoCont->getTPCTrack(gidTable[GTrackID::TPC]);
   const auto& trkITS = mRecoCont->getITSTrack(gidTable[GTrackID::ITS]);
   if (mDumpTrackPoints) {
     trackDataExtended = std::make_unique<TrackDataExtended>();
-    (*trackDataExtended).gid = (*mGIDs)[iSeed];
+    (*trackDataExtended).gid = mGIDs[iSeed];
     (*trackDataExtended).clIdx.setFirstEntry(mClRes.size());
     (*trackDataExtended).trkITS = trkITS;
     (*trackDataExtended).trkTPC = trkTPC;
@@ -193,15 +315,15 @@ void TrackInterpolation::interpolateTrack(int iSeed)
       (*trackDataExtended).clsITS.push_back(clsITS);
     }
   }
-  trackData.gid = (*mGIDs)[iSeed];
-  trackData.par = (*mSeeds)[iSeed];
-  auto& trkWork = (*mSeeds)[iSeed];
+  trackData.gid = mGIDs[iSeed];
+  trackData.par = mSeeds[iSeed];
+  auto& trkWork = mSeeds[iSeed];
   // reset the cache array (sufficient to set cluster available to zero)
   for (auto& elem : mCache) {
     elem.clAvailable = 0;
   }
   trackData.clIdx.setFirstEntry(mClRes.size()); // reference the first cluster residual belonging to this track
-  float clusterTimeBinOffset = (*mTrackTimes)[iSeed] / mTPCTimeBinMUS;
+  float clusterTimeBinOffset = mTrackTimes[iSeed] / mTPCTimeBinMUS;
 
   // store the TPC cluster positions in the cache
   for (int iCl = trkTPC.getNClusterReferences(); iCl--;) {
@@ -246,7 +368,7 @@ void TrackInterpolation::interpolateTrack(int iSeed)
     const auto& clTOF = mRecoCont->getTOFClusters()[gidTable[GTrackID::TOF]];
     if (mDumpTrackPoints) {
       (*trackDataExtended).clsTOF = clTOF;
-      (*trackDataExtended).matchTOF = mRecoCont->getTOFMatch((*mGIDs)[iSeed]);
+      (*trackDataExtended).matchTOF = mRecoCont->getTOFMatch(mGIDs[iSeed]);
     }
     const int clTOFSec = clTOF.getCount();
     const float clTOFAlpha = o2::math_utils::sector2Angle(clTOFSec);
@@ -409,8 +531,8 @@ void TrackInterpolation::interpolateTrack(int iSeed)
       (*trackDataExtended).clIdx.setEntries(nClValidated);
       mTrackDataExtended.push_back(std::move(*trackDataExtended));
     }
-    mGIDsSuccess.push_back((*mGIDs)[iSeed]);
-    mTrackDataCompact.emplace_back(mClRes.size() - nClValidated, nClValidated, (*mGIDs)[iSeed].getSource());
+    mGIDsSuccess.push_back(mGIDs[iSeed]);
+    mTrackDataCompact.emplace_back(mClRes.size() - nClValidated, nClValidated, mGIDs[iSeed].getSource());
   }
   if (mParams->writeUnfiltered) {
     TrackData trkDataTmp = trackData;
@@ -424,17 +546,18 @@ void TrackInterpolation::interpolateTrack(int iSeed)
 void TrackInterpolation::extrapolateTrack(int iSeed)
 {
   // extrapolate ITS-only track through TPC and store residuals to TPC clusters in the output vectors
-  const auto& gidTable = (*mGIDtables)[iSeed];
+  LOGP(debug, "Starting track extrapolation for GID {}", mGIDs[iSeed].asString());
+  const auto& gidTable = mGIDtables[iSeed];
   TrackData trackData;
   std::vector<TPCClusterResiduals> clusterResiduals;
   trackData.clIdx.setFirstEntry(mClRes.size());
   const auto& trkITS = mRecoCont->getITSTrack(gidTable[GTrackID::ITS]);
   const auto& trkTPC = mRecoCont->getTPCTrack(gidTable[GTrackID::TPC]);
-  trackData.gid = (*mGIDs)[iSeed];
-  trackData.par = (*mSeeds)[iSeed];
+  trackData.gid = mGIDs[iSeed];
+  trackData.par = mSeeds[iSeed];
 
-  auto& trkWork = (*mSeeds)[iSeed];
-  float clusterTimeBinOffset = (*mTrackTimes)[iSeed] / mTPCTimeBinMUS;
+  auto& trkWork = mSeeds[iSeed];
+  float clusterTimeBinOffset = mTrackTimes[iSeed] / mTPCTimeBinMUS;
   auto propagator = o2::base::Propagator::Instance();
   unsigned short rowPrev = 0; // used to calculate dRow of two consecutive cluster residuals
   unsigned short nMeasurements = 0;
@@ -494,8 +617,8 @@ void TrackInterpolation::extrapolateTrack(int iSeed)
     }
     trackData.clIdx.setEntries(nClValidated);
     mTrackData.push_back(std::move(trackData));
-    mGIDsSuccess.push_back((*mGIDs)[iSeed]);
-    mTrackDataCompact.emplace_back(mClRes.size() - nClValidated, nClValidated, (*mGIDs)[iSeed].getSource());
+    mGIDsSuccess.push_back(mGIDs[iSeed]);
+    mTrackDataCompact.emplace_back(mClRes.size() - nClValidated, nClValidated, mGIDs[iSeed].getSource());
   }
   if (mParams->writeUnfiltered) {
     TrackData trkDataTmp = trackData;
@@ -881,6 +1004,13 @@ void TrackInterpolation::reset()
   mTrackDataUnfiltered.clear();
   mClResUnfiltered.clear();
   mGIDsSuccess.clear();
+  for (auto& vec : mTrackIndices) {
+    vec.clear();
+  }
+  mGIDs.clear();
+  mGIDtables.clear();
+  mTrackTimes.clear();
+  mSeeds.clear();
 }
 
 //______________________________________________

--- a/Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
+++ b/Detectors/TRD/workflow/src/TRDGlobalTrackingSpec.cxx
@@ -525,9 +525,6 @@ bool TRDGlobalTracking::refitITSTPCTRDTrack(TrackTRD& trk, float timeTRD, o2::gl
   int nCl = -1, clEntry = -1, nClRefit = 0, clRefs[14];
   float chi2Out = 0, timeZErr = 0.;
   bool pileUpOn = trk.hasPileUpInfo(); // distance to farthest collision within the pileup integration time is set
-  if (pileUpOn) {
-    timeTRD += trk.getPileUpTimeShiftMUS(); // shift to average pileup position
-  }
   auto geom = o2::its::GeometryTGeo::Instance();
   auto matCorr = o2::base::Propagator::MatCorrType(mRec->GetParam().rec.trd.matCorrType);
   if (detRefs[GTrackID::ITS].isIndexSet()) { // this is ITS track
@@ -656,10 +653,7 @@ bool TRDGlobalTracking::refitTPCTRDTrack(TrackTRD& trk, float timeTRD, o2::globa
   auto detRefs = recoCont->getSingleDetectorRefs(trk.getRefGlobalTrackId());
   outerParam = trk;
   float chi2Out = 0, timeZErr = 0.;
-  bool pileUpOn = trk.hasPileUpInfo(); // distance to farthest collision within the pileup integration time is set
-  if (pileUpOn) {
-    timeTRD += trk.getPileUpTimeShiftMUS(); // shift to average pileup position
-  }
+  bool pileUpOn = trk.hasPileUpInfo();                                                                                                                                                           // distance to farthest collision within the pileup integration time is set
   int retVal = mTPCRefitter->RefitTrackAsTrackParCov(outerParam, mTPCTracksArray[detRefs[GTrackID::TPC]].getClusterRef(), (timeTRD + mTPCTDriftOffset) * mTPCTBinMUSInv, &chi2Out, true, false); // outward refit
   if (retVal < 0) {
     LOG(debug) << "TPC refit outwards failed";

--- a/prodtests/full-system-test/calib-workflow.sh
+++ b/prodtests/full-system-test/calib-workflow.sh
@@ -23,7 +23,7 @@ fi
 if [[ "${CALIB_TPC_SCDCALIB_SENDTRKDATA:-}" == "1" ]]; then ENABLE_TRKDATA_OUTPUT="--send-track-data"; else ENABLE_TRKDATA_OUTPUT=""; fi
 
 # specific calibration workflows
-if [[ $CALIB_TPC_SCDCALIB == 1 ]]; then add_W o2-tpc-scdcalib-interpolation-workflow "${CALIB_TPC_SCDCALIB_SLOTLENGTH:+"--sec-per-slot $CALIB_TPC_SCDCALIB_SLOTLENGTH"} $ENABLE_TRKDATA_OUTPUT $DISABLE_ROOT_OUTPUT --disable-root-input --pipeline $(get_N tpc-track-interpolation TPC REST)"; fi
+if [[ $CALIB_TPC_SCDCALIB == 1 ]]; then add_W o2-tpc-scdcalib-interpolation-workflow "--vtx-sources $VERTEX_TRACK_MATCHING_SOURCES ${CALIB_TPC_SCDCALIB_SLOTLENGTH:+"--sec-per-slot $CALIB_TPC_SCDCALIB_SLOTLENGTH"} $ENABLE_TRKDATA_OUTPUT $DISABLE_ROOT_OUTPUT --disable-root-input --pipeline $(get_N tpc-track-interpolation TPC REST)"; fi
 if [[ $CALIB_TPC_TIMEGAIN == 1 ]]; then
   : ${SCALEEVENTS_TPC_TIMEGAIN:=40}
   : ${SCALETRACKS_TPC_TIMEGAIN:=1000}

